truffleruby/src/main/java/org/truffleruby/language/dispatch/CachedForeignDispatchNode.java

```diff
@@ -29,7 +29,7 @@ public CachedForeignDispatchNode(DispatchNode next, Object cachedName) {
 
     @Override
     protected boolean guard(Object methodName, Object receiver) {
-        return guardName(methodName) && (receiver instanceof TruffleObject);
+        return guardName(methodName) && !(receiver instanceof DynamicObject) && (receiver instanceof TruffleObject);
```

Comment:
This is not quite enough, JS also uses DynamicObject.
You should use isForeignObject: https://github.com/graalvm/truffleruby/blob/05efbb977df5fa0cc16ed4e11adbd6a31ea8a15d/truffleruby/src/main/java/org/truffleruby/language/RubyGuards.java#L262